File: [46] Permutations/46.permutations.java
import java.awt.List;
import java.util.ArrayList;

/*
 * @lc app=leetcode id=46 lang=java
 *
 * [46] Permutations
 */

// @lc code=start
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        DFS(nums, res, new ArrayList<Integer>());
        return res;
    }

    public void DFS(int[] nums, List<List<Integer>> res, List<Integer> temp_list) {
        if (temp_list.size() == nums.length) {
            res.add(new ArrayList<>(temp_list));
            return;
        } else {
            for (int i = 0; i < nums.length; i++) {
                if (temp_list.contains(nums[i])) continue;
                temp_list.add(nums[i]);
                DFS(nums, res, temp_list);
                temp_list.remove(temp_list.size()-1);
            }
        }
    }
}
// @lc code=end

File: [46] Permutations/46.permutations.cpp
/*
 * @lc app=leetcode id=46 lang=cpp
 *
 * [46] Permutations
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        // this method swap nums position
        vector<vector<int>> res;
        DFS(nums, res, 0);
        return res;
    }

    // permute num[start..end]
    // invariant: num[0..start-1] have been fixed/permuted
    void DFS(vector<int>& nums, vector<vector<int>>& res, int start) {
        if (start == nums.size()) {
            res.emplace_back(nums);
            return;
        } else {
            for (int i = start; i < nums.size(); i++) {
                swap(nums[start], nums[i]);
                DFS(nums, res, start+1);
                // reset
                swap(nums[start], nums[i]);
            }
        }
    }
};
// @lc code=end

